### ISSUE TITLE
[JENKINS-43903] email-ext jelly template slave node absolute path no more handled

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/email-ext-plugin#1225]
<p>The email-ext lost a feature between </p>

<p><b>working</b>: Jenkins 2.51, email-ext 2.56</p>

<p><b>not working</b>: Jenkins 2.54, email-ext 2.57-2</p>

<p>The template cannot reside on the slave node anymore, the template file is not found anymore with absolute path. The following pipeline code on inside a Windows slave node:</p>

<p> </p>

<pre>
def email_template_file = pwd([tmp: <span class="code-keyword">true</span>]) + <span class="code-quote">"\\email_template_file.jelly"</span>;
writeFile([file: email_template_file, text: content_template]);
emailext body: <span class="code-quote">'${JELLY_SCRIPT,template=<span class="code-quote">"'</span> + email_template_file.replaceAll(<span class="code-quote">'[\\\\]'</span>,<span class="code-quote">'/'</span>) + <span class="code-quote">'"</span>}'</span></pre>
<p> The email got send but the template is no more found, now give the following results into the email body:</p>

<pre>
Jelly file [c:/Jenkins/workspace/CAD_CPP/CAD_CPP_ContinuousBuild@​tmp/email_template_file.jelly] was not found in$JENKINS_HOME/email-templates.
</pre>
<p>The workaround is to save the email template on the master (hope no other does the same at the same time or use guid into the filename):</p>

<p> </p>

<pre>

def email_template_file = <span class="code-quote">"testid.html"</span>;
node(<span class="code-quote">'master'</span>)
{
writeFile([file: <span class="code-quote">"${JENKINS_HOME}/email-templates/"</span> + email_template_file, text: content_template]);
emailext body: <span class="code-quote">'${JELLY_SCRIPT,template=<span class="code-quote">"'</span> + email_template_file.replaceAll(<span class="code-quote">'[\\\\]'</span>,<span class="code-quote">'/'</span>) + <span class="code-quote">'"</span>}'</span>, subject: <span class="code-quote">'test subject'</span>, to: <span class="code-quote">'test@​test.com'</span>, replyTo: <span class="code-quote">'jenkins@​test.com'</span>, mimeType: <span class="code-quote">'text/html'</span>;

}
</pre>
<p>This is used when you need to determine which section of the email template must be used at run time and based on run results.  The pipeline script concat email template section together and then use the results as the final email template.</p>

<p>Now it's locking the template path to a relative path on the master and on top of it, it enforce the ${JENKINS_HOME}/email-templates folder.</p>

<p> </p>

<p> </p>

---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jerome_godbout">jerome_godbout</a>, imported from: <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-43903" target="_blank">email-ext jelly template slave node absolute path no more handled</a></i></summary>
<i><ul>
<li><b>assignee</b>: <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/davidvanlaatum.png" title="davidvanlaatum's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=davidvanlaatum">davidvanlaatum</a>
<li><b>status</b>: Closed
<li><b>priority</b>: Minor
<li><b>component(s)</b>: email-ext-plugin, emailext-template-plugin
<li><b>label(s)</b>: email-ext, node, path, pipeline
<li><b>resolution</b>: Won't Fix
<li><b>resolved</b>: 2017-08-15T08:41:58+00:00
<li><b>votes</b>: 0
<li><b>watchers</b>: 2
<li><b>imported</b>: 20251211-193728
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>The email-ext lost a feature between </p>

<p><b>working</b>: Jenkins 2.51, email-ext 2.56</p>

<p><b>not working</b>: Jenkins 2.54, email-ext 2.57-2</p>

<p>The template cannot reside on the slave node anymore, the template file is not found anymore with absolute path. The following pipeline code on inside a Windows slave node:</p>

<p> </p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">def email_template_file = pwd([tmp: <span class="code-keyword">true</span>]) + <span class="code-quote">"\\email_template_file.jelly"</span>;
writeFile([file: email_template_file, text: content_template]);
emailext body: <span class="code-quote">'${JELLY_SCRIPT,template=<span class="code-quote">"'</span> + email_template_file.replaceAll(<span class="code-quote">'[\\\\]'</span>,<span class="code-quote">'/'</span>) + <span class="code-quote">'"</span>}'</span></pre>
</div></div>
<p> The email got send but the template is no more found, now give the following results into the email body:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">Jelly file [c:/Jenkins/workspace/CAD_CPP/CAD_CPP_ContinuousBuild@tmp/email_template_file.jelly] was not found in$JENKINS_HOME/email-templates.
</pre>
</div></div>
<p>The workaround is to save the email template on the master (hope no other does the same at the same time or use guid into the filename):</p>

<p> </p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">
def email_template_file = <span class="code-quote">"testid.html"</span>;
node(<span class="code-quote">'master'</span>)
{
writeFile([file: <span class="code-quote">"${JENKINS_HOME}/email-templates/"</span> + email_template_file, text: content_template]);
emailext body: <span class="code-quote">'${JELLY_SCRIPT,template=<span class="code-quote">"'</span> + email_template_file.replaceAll(<span class="code-quote">'[\\\\]'</span>,<span class="code-quote">'/'</span>) + <span class="code-quote">'"</span>}'</span>, subject: <span class="code-quote">'test subject'</span>, to: <span class="code-quote">'test@test.com'</span>, replyTo: <span class="code-quote">'jenkins@test.com'</span>, mimeType: <span class="code-quote">'text/html'</span>;

}
</pre>
</div></div>
<p>This is used when you need to determine which section of the email template must be used at run time and based on run results.  The pipeline script concat email template section together and then use the results as the final email template.</p>

<p>Now it's locking the template path to a relative path on the master and on top of it, it enforce the ${JENKINS_HOME}/email-templates folder.</p>

<p> </p>

<p> </p></pre>
</details>
</details>
<details><summary><i>environment</i></summary>

```
Linux Master<br/>
Windows Node
```
</details>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/davidvanlaatum.png" title="davidvanlaatum's avatar" /> davidvanlaatum:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-43903?focusedId=310442&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-310442">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>This was changed for security reasons </p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java"><span class="code-keyword">if</span>(!scriptFile.exists() || !AbstractEvalContent.isChildOf(<span class="code-keyword">new</span> FilePath(scriptFile), <span class="code-keyword">new</span> FilePath(scriptsFolder))) {</pre>
</div></div>
<p>If it fails that check it gives not found</p>
</pre>
</details>

</ul>
</details>

<p>This was changed for security reasons </p>

<pre>
<span class="code-keyword">if</span>(!scriptFile.exists() || !AbstractEvalContent.isChildOf(<span class="code-keyword">new</span> FilePath(scriptFile), <span class="code-keyword">new</span> FilePath(scriptsFolder))) {</pre>
<p>If it fails that check it gives not found</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>jerome_godbout:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-43903?focusedId=310458&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-310458">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>That's sad that this feature is gone, and I guess for the name of security again. How one generate a dynamic template on node and manage to send it properly now? right now I have to stash the template generated on a node, check for a lock file and unstash on the master, send the email and release the lock file on the master. It doesn't feel like a funny Jenkinsfile sequence only to send an email. I still can live with it, but my guess is that newcomer will have trouble with it if they venture with dynamic template (email template content depends on the build sequence used into the jenkinsfile and the result of it).</p>
</pre>
</details>

</ul>
</details>

<p>That's sad that this feature is gone, and I guess for the name of security again. How one generate a dynamic template on node and manage to send it properly now? right now I have to stash the template generated on a node, check for a lock file and unstash on the master, send the email and release the lock file on the master. It doesn't feel like a funny Jenkinsfile sequence only to send an email. I still can live with it, but my guess is that newcomer will have trouble with it if they venture with dynamic template (email template content depends on the build sequence used into the jenkinsfile and the result of it).</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/davidvanlaatum.png" title="davidvanlaatum's avatar" /> davidvanlaatum:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-43903?focusedId=310594&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-310594">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I know I wish it wasn't the case but as one of the more popular plugins I have to address security issues when they are raised. There is a lock pipeline plugin that can simplify your steps there by wrapping the copy file and send email in a lock block that will wait on a specified global lock. Alternatively you can generate the html of the email in the pipeline itself or in a pipeline library. These use the groovy sandboxing and as such are not as open to security problems.</p>
</pre>
</details>

</ul>
</details>

<p>I know I wish it wasn't the case but as one of the more popular plugins I have to address security issues when they are raised. There is a lock pipeline plugin that can simplify your steps there by wrapping the copy file and send email in a lock block that will wait on a specified global lock. Alternatively you can generate the html of the email in the pipeline itself or in a pipeline library. These use the groovy sandboxing and as such are not as open to security problems.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>vipivoxa:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-43903?focusedId=316817&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-316817">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/> higher security introduces kind of less security, now e.g. email templates, used to managed by their own projects (have multi project environment here using one jenkins instance with many nodes for many different projects) stored on node level, have to be applied via jenkins admin. And in reality admin is not going through every single line of email template, some of them are pretty cumbersome ....all admin does - just applies script by request.</p>

<p>Also this additional security reason, introduces more manual steps to the process, which we kind of trying to reduce here - before, auto builds might have used templates from a source control dir, checked out and operated, obviously, on node level, now it is not the case.</p>
</pre>
</details>

</ul>
</details>

<p><img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/> higher security introduces kind of less security, now e.g. email templates, used to managed by their own projects (have multi project environment here using one jenkins instance with many nodes for many different projects) stored on node level, have to be applied via jenkins admin. And in reality admin is not going through every single line of email template, some of them are pretty cumbersome ....all admin does - just applies script by request.</p>

<p>Also this additional security reason, introduces more manual steps to the process, which we kind of trying to reduce here - before, auto builds might have used templates from a source control dir, checked out and operated, obviously, on node level, now it is not the case.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/davidvanlaatum.png" title="davidvanlaatum's avatar" /> davidvanlaatum:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-43903?focusedId=316871&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-316871">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I agree but I have to support security, if someone wants to implement a way of keeping the templates in a SCM like we can do with pipeline libs now I will accept a pull</p>
</pre>
</details>

</ul>
</details>

<p>I agree but I have to support security, if someone wants to implement a way of keeping the templates in a SCM like we can do with pipeline libs now I will accept a pull</p>
